### PR TITLE
Add submodule and tests for private clusters

### PR DESCRIPTION
**Pull request by Jberlinsky**

### CODE REVIEW

File: autogen/README.md [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+4. If you are using a Shared VPC, the APIs must also be activated on the Shared VPC host project and your service account needs the proper permissions there.
+
+The [project factory](https://github.com/terraform-google-modules/terraform-google-project-factory) can be used to provision projects with the correct APIs active and the necessary Shared VPC connections.
+
+### Software Dependencies
+### Kubectl
+- [kubectl](https://github.com/kubernetes/kubernetes/releases) 1.9.x
+### Terraform plugins
+- [Terraform](https://www.terraform.io/downloads.html) 0.10.x
+- [terraform-provider-google](https://github.com/terraform-providers/terraform-provider-google) plugin {% if private_cluster %}beta{% else %}v1.8.0{% endif %}
```
Comment (aaron-lane) [inline autogen/README.md]:
```suggestion
{% if private_cluster %}
- [terraform-provider-google-beta](https://github.com/terraform-providers/terraform-provider-google-beta) v1.20.0
{% else %}
- [terraform-provider-google](https://github.com/terraform-providers/terraform-provider-google) v1.8.0
{% endif %}
```

File: autogen/README.md [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+### Enable APIs
+In order to operate with the Service Account you must activate the following APIs on the project where the Service Account was created:
+
+- Compute Engine API - compute.googleapis.com
+- Kubernetes Engine API - container.googleapis.com
+
+## Install
+
+### Terraform
+Be sure you have the correct Terraform version (0.10.x), you can choose the binary here:
```
Comment (aaron-lane) [inline autogen/README.md]: This should be moved under `### Software Dependencies` and it should reference 0.11.x.

File: autogen/README.md [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+
+1. Terraform and kubectl are [installed](#software-dependencies) on the machine where Terraform is executed.
+2. The Service Account you execute the module with has the right [permissions](#iam-roles).
+3. The Compute Engine and Kubernetes Engine APIs are [active](#enable-apis) on the project you will launch the cluster in.
+4. If you are using a Shared VPC, the APIs must also be activated on the Shared VPC host project and your service account needs the proper permissions there.
+
+The [project factory](https://github.com/terraform-google-modules/terraform-google-project-factory) can be used to provision projects with the correct APIs active and the necessary Shared VPC connections.
+
+### Software Dependencies
+### Kubectl
```
Comment (aaron-lane) [inline autogen/README.md]:
```suggestion
#### Kubectl
```

File: autogen/README.md [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+- [kubectl](https://github.com/kubernetes/kubernetes/releases) 1.9.x
+### Terraform plugins
```
Comment (aaron-lane) [inline autogen/README.md]:
```suggestion
#### Terraform Plugins
```

File: examples/simple_regional_private/README.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# Simple Regional Cluster
+
+This example illustrates how to create a simple private cluster.
+
+[^]: (autogen_docs_start)
+
```
Comment (aaron-lane) [inline examples/simple_regional_private/README.md]: Documentation needs to be generated.

File: helpers/generate_modules/generate_modules.py [hunk 6]
```diff
@@ -21,29 +21,63 @@
...
+
+MODULES = [
+    Module("./", {
+        'private_cluster': False,
+    }),
+    Module("./modules/private-cluster", {
+        'private_cluster': True,
+    }),
+]
+DEVNULL_FILE = open(os.devnull, 'w')
```
Comment (aaron-lane) [inline helpers/generate_modules/generate_modules.py]: Does this need to be explicitly closed?

**Comment by Jberlinsky** [comment]
@adrienthebo Thanks for checking in, and apologies for the delay. This PR is now in a workable state, and is ready for review.

This is indeed a hefty change, due to the consolidation of logic in `autogen/`. The particularly relevant bits of this PR are the changes in `autogen/cluster_regional.tf` and `autogen/cluster_zonal.tf`, as they template out the major differences between the public and private cluster modules. I also added two test cases (suffixed with `_private`) to test such clusters.

**Comment by Jberlinsky** [comment]
Thanks @adrienthebo!

@aaron-lane could you kindly dismiss your review (or re-review if you're so inclined) so I can merge this?

**Comment by adrienthebo** [comment]
Verified that this change does not cause any Terraform state changes when upgraded from master.

**Pull request merged**